src/datadog/baggage.cpp

```diff
@@ -0,0 +1,250 @@
+          goto value;
+        }
+      } break;
+
+      case state::value: {
+      value:
+        if (c == ',') {
+          tmp_end = i;
+          goto consume_value;
+        } else if (is_whitespace(c)) {
+          tmp_end = i;
+          internal_state = state::trailing_spaces_value;
+        } else if (!is_allowed_value_char(c)) {
+          return Baggage::Error{Baggage::Error::MALFORMED_BAGGAGE_HEADER, i};
+        }
+      } break;
+
+      case state::trailing_spaces_value: {
+        if (c == ',') {
+        consume_value:
+          size_t count = tmp_end - beg;
+          if (count < 1)
+            return Baggage::Error{Baggage::Error::MALFORMED_BAGGAGE_HEADER,
+                                  tmp_end};
+
+          value = StringView{input.data() + beg, count};
+          result.emplace(std::string(key), std::string(value));
+          beg = i;
+          tmp_end = i;
+          internal_state = state::leading_spaces_key;
+        } else if (!is_whitespace(c)) {
+          return Baggage::Error{Baggage::Error::MALFORMED_BAGGAGE_HEADER, i};
+        }
+      }
+    }
+  }
+
+  if (internal_state == state::value) {
+    value = StringView{input.data() + beg, end - beg};
+    result.emplace(std::string(key), std::string(value));
+  } else if (internal_state == state::trailing_spaces_value) {
+    value = StringView{input.data() + beg, tmp_end - beg};
+    result.emplace(std::string(key), std::string(value));
+  } else {
+    return Baggage::Error{Baggage::Error::MALFORMED_BAGGAGE_HEADER, end};
+  }
+
+  return result;
+}
+
+}  // namespace
+
+Baggage::Baggage(size_t max_capacity) : max_capacity_(max_capacity) {}
+
+Baggage::Baggage(std::unordered_map<std::string, std::string> baggage,
+                 size_t max_capacity)
+    : max_capacity_(max_capacity), baggage_(std::move(baggage)) {}
+
+Optional<StringView> Baggage::get(StringView key) const {
+  auto it = baggage_.find(std::string(key));
+  if (it == baggage_.cend()) return nullopt;
+
+  return it->second;
+}
+
+bool Baggage::set(std::string key, std::string value) {
+  if (baggage_.size() == max_capacity_) return false;
+
+  baggage_[key] = value;
+  return true;
+}
+
+void Baggage::remove(StringView key) { baggage_.erase(std::string(key)); }
+
+void Baggage::clear() { baggage_.clear(); }
+
+size_t Baggage::size() const { return baggage_.size(); }
+
+bool Baggage::empty() const { return baggage_.empty(); }
+
+bool Baggage::contains(StringView key) const {
+  auto found = baggage_.find(std::string(key));
+  return found != baggage_.cend();
+}
+
+void Baggage::visit(std::function<void(StringView, StringView)>&& visitor) {
+  for (const auto& [key, value] : baggage_) {
+    visitor(key, value);
+  }
+}
+
+Expected<void> Baggage::inject(DictWriter& writer, const Options& opts) const {
+  if (baggage_.empty()) return {};
+  if (baggage_.size() > opts.max_items)
+    return datadog::tracing::Error{
+        datadog::tracing::Error::Code::BAGGAGE_MAXIMUM_BYTES_REACHED, ""};
```

Comment:
If we keep this behavior after the discussion posted above, then:

```suggestion
        datadog::tracing::Error::Code::MAXIMUM_CAPACITY_REACHED, ""};
```